### PR TITLE
pybind/mgr: add support for common sqlite3 databases

### PR DESCRIPTION
**Pull request by batrick**
And update "devicehealth" pool to use it.

The one nit in this PR is that the mgr normally can't delete the "device_health_metrics" pool because `mon_allow_pool_delete` is normally false. I'm not yet sure how we want to address this; I'm open to suggestions.

I've manually imported the LRC device_health_metrics to test loading a legacy pool.

### CODE REVIEW

**Comment by batrick** [comment]
jenkins test api

**Comment by runsisi** [comment]
impllicit pools, e.g. `devicehealth`, is a very weird design.. our customs never agree with the community on these implicit things. imho, we should store all those infos into our own db, like mon db or something else.

**Comment by runsisi** [comment]
impllicit pools, e.g. `devicehealth`, is a very weird design.. our customs never agree with the community on these implicit things. imho, we should store all those infos into our own db, like mon db or something else.

**Comment by batrick** [comment]
> impllicit pools, e.g. `devicehealth`, is a very weird design.. our customs never agree with the community on these implicit things. imho, we should store all those infos into our own db, like mon db or something else.

It's a direction we've been going already. This PR actually improves on that a bit because we will be discouraging the creation of a unique pool for each mgr module.

The mgr is intended to offload the monitors so we should be avoiding the `config-key` changes that devicehealth was doing (`last_scrape`).

**Comment by batrick** [comment]
> The one nit in this PR is that the mgr normally can't delete the "device_health_metrics" pool because `mon_allow_pool_delete` is normally false. I'm not yet sure how we want to address this; I'm open to suggestions.

One option I kinda like is to rename `device_health_metrics` to `mgr`. Then after injesting the old objects, delete them. Thoughts?

**Comment by jdurgin** [comment]
> > impllicit pools, e.g. `devicehealth`, is a very weird design.. our customs never agree with the community on these implicit things. imho, we should store all those infos into our own db, like mon db or something else.
> 
> It's a direction we've been going already. This PR actually improves on that a bit because we will be discouraging the creation of a unique pool for each mgr module.
> 
> The mgr is intended to offload the monitors so we should be avoiding the `config-key` changes that devicehealth was doing (`last_scrape`).

Yes, using config-key isn't scalable for high update rates or large data sets, which becomes an issue with some of the modules already, hence using pools.

**Comment by jdurgin** [comment]
> > The one nit in this PR is that the mgr normally can't delete the "device_health_metrics" pool because `mon_allow_pool_delete` is normally false. I'm not yet sure how we want to address this; I'm open to suggestions.
> 
> One option I kinda like is to rename `device_health_metrics` to `mgr`. Then after injesting the old objects, delete them. Thoughts?

I think renaming would be better. You could avoid a potentially very long bulk ingest and rely on pruning old data to get rid of it instead.

**Comment by jdurgin** [comment]
> > > The one nit in this PR is that the mgr normally can't delete the "device_health_metrics" pool because `mon_allow_pool_delete` is normally false. I'm not yet sure how we want to address this; I'm open to suggestions.
> > 
> > 
> > One option I kinda like is to rename `device_health_metrics` to `mgr`. Then after injesting the old objects, delete them. Thoughts?
> 
> I think renaming would be better. You could avoid a potentially very long bulk ingest and rely on pruning old data to get rid of it instead.

You could still do ingest into the new format, I think you'll want to break it up though so it can't overwhelm the cluster. we default to saving 180 days, which could get large for large clusters. You could convert a few at a time in the background.

**Comment by batrick** [comment]
jenkins test dashboard

**Comment by batrick** [comment]
@jdurgin what do you think of the new version? Only downside of the incremental ingest is that the database will be out of date briefly. That's fixable but it'd be laborious to code up.

**Comment by batrick** [comment]
Other major change I made is that the "mgr" pool is now "__mgr" to reduce the potential for conflicts with user pools. I figure it's the right time to indicate that some pool names are reserved (beginning with "__") which I've noted in the docs.

**Comment by liewegas** [comment]
> Other major change I made is that the `mgr` pool is now `__mrg` to reduce the potential for conflicts with user pools. I figure it's the right time to indicate that some pool names are reserved (beginning with `__`) which I've noted in the docs.

We already have a ``.rgw.root`` pool; i wonder if we should use ``.`` as the "reserved" prefix char?

**Comment by batrick** [comment]
> > Other major change I made is that the `mgr` pool is now `__mrg` to reduce the potential for conflicts with user pools. I figure it's the right time to indicate that some pool names are reserved (beginning with `__`) which I've noted in the docs.
> 
> We already have a `.rgw.root` pool; i wonder if we should use `.` as the "reserved" prefix char?

Sounds good to me. I'll make the change.

File: src/pybind/mgr/mgr_module.py [hunk 1]
```diff
@@ -920,6 +940,184 @@ def cluster_log(self, channel: str, priority: ClusterLogPrio, message: str) -> N
...
 
+    def pool_exists(self, name: str) -> bool:
+        pools = [p['pool_name'] for p in self.get('osd_map')['pools']]
+        return name in pools
+
+    def have_enough_osds(self) -> bool:
+        # wait until we have enough OSDs to allow the pool to be healthy
+        up = 0
+        for osd in self.get("osd_map")["osds"]:
+            if osd["up"]:
```
Comment (jdurgin) [inline src/pybind/mgr/mgr_module.py:951]: you probably want up + in, to ensure they're available for data placement as well as running

File: src/pybind/mgr/mgr_module.py [hunk 2]
```diff
@@ -920,6 +940,184 @@ def cluster_log(self, channel: str, priority: ClusterLogPrio, message: str) -> N
     def version(self) -> str:
         return self._version
 
+    def pool_exists(self, name: str) -> bool:
+        pools = [p['pool_name'] for p in self.get('osd_map')['pools']]
```
Comment (jdurgin) [inline src/pybind/mgr/mgr_module.py:944]: maybe a future todo, copying/jsonifying the osdmap on every access to self.db could get expensive for large clusters. it'd be better to use a get_pools() or pool_exists() c++ method - could be added to mgr/PyOSDMap.cc so you could do self.get_osdmap().get_pools() or something similar

File: src/pybind/mgr/devicehealth/module.py [hunk 3]
```diff
@@ -243,70 +279,87 @@ def config_notify(self) -> None:
...
-            'prefix': 'osd pool application enable',
-            'format': 'json',
-            'pool': self.pool_name,
-            'app': 'mgr_devicehealth',
-        }), '')
-        r, outb, outs = result.wait()
-        assert r == 0
+    def check_legacy_pool(self) -> bool:
+        try:
+            # 'device_health_metrics' is automatically renamed '__mgr' in
```
Comment (jdurgin) [inline src/pybind/mgr/devicehealth/module.py:316]: s/__/./

**Review by jdurgin — COMMENTED** [review]
looks good in general! the incremental ingest seems like a good approach to me

File: src/pybind/mgr/mgr_module.py (hunk 2 above)
Comment (batrick) [reply src/pybind/mgr/mgr_module.py:944]: This only happens if `self._db is None` (usually the first access by each module).

File: src/pybind/mgr/devicehealth/module.py (hunk 3 above)
Comment (batrick) [reply src/pybind/mgr/devicehealth/module.py]: Fixed, thanks

File: src/pybind/mgr/mgr_module.py (hunk 1 above)
Comment (batrick) [reply src/pybind/mgr/mgr_module.py]: done

File: src/pybind/mgr/mgr_module.py [hunk 4]
```diff
@@ -920,6 +940,184 @@ def cluster_log(self, channel: str, priority: ClusterLogPrio, message: str) -> N
...
+        db.execute('PRAGMA CACHE_SIZE = 64')
+        db.row_factory = sqlite3.Row
+        self.load_schema(db)
+
+    def open_db(self) -> Optional[sqlite3.Connection]:
+        if not self.pool_exists(self.MGR_POOL_NAME):
+            if not self.have_enough_osds():
+                return None
+            self.create_mgr_pool()
+        uri = f"file:///{self.MGR_POOL_NAME}:{self.module_name}/main.db?vfs=ceph";
```
Comment (jdurgin) [inline src/pybind/mgr/mgr_module.py:1097]: @yaarith I hadn't noticed before there's actually already a namespace here - the module name - so we should be good for future proofing

**Comment by batrick** [comment]
resolved conflicts with master

File: src/pybind/mgr/mgr_module.py [hunk 5]
```diff
@@ -920,6 +940,184 @@ def cluster_log(self, channel: str, priority: ClusterLogPrio, message: str) -> N
...
+        result = CommandResult('')
+        self.send_command(result, 'mon', '', json.dumps({
+            'prefix': 'osd pool application enable',
+            'format': 'json',
+            'pool': name,
+            'app': app,
+        }), '')
+        r, outb, outs = result.wait()
+        if r is not 0:
+            raise RuntimeError(f"cannot set \"{name}\" pool application")
```
Comment (varshar16) [inline src/pybind/mgr/mgr_module.py:992]: Instead  of `send_command` use `check_mon_command`. It checks the return value

Comment (batrick) [reply src/pybind/mgr/mgr_module.py]: Thanks for pointing that out -- fixed!

**Review by jdurgin — CHANGES_REQUESTED** [review]
Could you add a test to verify the migration path works? Maybe something in an upgrade suite like upgrade/octopus-x that verifies that we have device health data before upgrading to master

**Comment by batrick** [comment]
> Could you add a test to verify the migration path works? Maybe something in an upgrade suite like upgrade/octopus-x that verifies that we have device health data before upgrading to master

Okay, I've added a primitive test. I don't really like that it's done for every upgrade test but couldn't find a better place for it. Any other suggestions?

**Comment by batrick** [comment]
jenkins test docs

**Comment by batrick** [comment]
jenkins test docs

**Comment by batrick** [comment]
jenkins test dashboard

**Review by jdurgin — APPROVED** [review]

**Comment by batrick** [comment]
Fixed a YAML syntax error. Testing here: https://pulpito.ceph.com/pdonnell-2021-04-20_02:14:43-upgrade:octopus-x:parallel-wip-pdonnell-testing-20210419.204944-distro-basic-smithi/

**Comment by tchaikov** [comment]
jenkins test api

**Comment by batrick** [comment]
Added `--yes-i-really-mean-it` to `appify_pool`.

**Comment by jdurgin** [comment]
jenkins test docs

**Comment by batrick** [comment]
fixed a small bug in the new upgrade test

New QA run here: https://pulpito.ceph.com/?branch=wip-pdonnell-testing-20210420.204741

New run with Ubuntu (slow builds) will be scheduled in 2h: https://pulpito.ceph.com/?branch=wip-pdonnell-testing-20210420.204741

**Comment by batrick** [comment]
jenkins test docs

File: src/pybind/mgr/devicehealth/module.py [hunk 6]
```diff
@@ -278,15 +279,78 @@ def config_notify(self) -> None:
...
+            ioctx = self.rados.open_ioctx(self.MGR_POOL_NAME)
+        except rados.ObjectNotFound:
+            return True
+        if not ioctx:
+            return True
+
+        done = False
+        with ioctx:
+            with self._db_lock:
+                with self.db:
```
Comment (tchaikov) [inline src/pybind/mgr/devicehealth/module.py:327]: nit, could use a more compacted form of `with` statement, like
```py
with ioctx, self._db_lock, self.db:
    count = 0
    # ...
```

see https://docs.python.org/3/reference/compound_stmts.html#the-with-statement. where multiple `with_item`s are allowed.

Comment (batrick) [reply src/pybind/mgr/devicehealth/module.py:327]: What a relief, I'll change that.

**Comment by batrick** [comment]
jenkins test dashboard

**Comment by batrick** [comment]
jenkins test dashboard

**Comment by batrick** [comment]
New run: https://pulpito.ceph.com/pdonnell-2021-04-21_22:58:01-upgrade:octopus-x:parallel-wip-pdonnell-testing-20210421.204942-distro-basic-smithi/

File: src/pybind/mgr/devicehealth/module.py [hunk 7]
```diff
@@ -489,37 +525,39 @@ def put_device_metrics(self, ioctx: rados.Ioctx, devid: str, data: Any) -> None:
...
-        with ioctx:
-            with rados.ReadOpCtx() as op:
-                omap_iter, ret = ioctx.get_omap_vals(op, min_sample or '', sample or '',
-                                                     MAX_SAMPLES)  # fixme
-                assert ret == 0
+
+        SQL = """
+        SELECT time, raw_smart
+            FROM DeviceHealthMetrics
+            WHERE devid = ? AND (time = ? OR ? <= time)
```
Comment (yaarith) [inline src/pybind/mgr/devicehealth/module.py:542]: I think that if `min_sample` is set, then we're looking for records that are newer than it;
therefore:
`? >= time`

Comment (batrick) [reply src/pybind/mgr/devicehealth/module.py:542]: That would select all records older than `time`. I don't think so?

Comment (yaarith) [reply src/pybind/mgr/devicehealth/module.py:542]: sorry, I read it the other way around, you're right!

File: src/pybind/mgr/mgr_module.py [hunk 8]
```diff
@@ -920,6 +940,174 @@ def cluster_log(self, channel: str, priority: ClusterLogPrio, message: str) -> N
...
+
+    def set_kv(self, key: str, value: Any) -> None:
+        SQL = "INSERT OR REPLACE INTO MgrModuleKV (key, value) VALUES (?, ?);"
+
+        assert key[:2] != "__"
+
+        self.log.debug(f"set_kv('{key}', '{value}')")
+
+        with self._db_lock, self.db:
+                self.db.execute(SQL, (key, value))
```
Comment (tchaikov) [inline src/pybind/mgr/mgr_module.py:1024]: wrong indent.

File: src/pybind/mgr/mgr_module.py [hunk 9]
```diff
@@ -920,6 +940,174 @@ def cluster_log(self, channel: str, priority: ClusterLogPrio, message: str) -> N
...
+        self.configure_db(db)
+        return db
+
+    def db_ready(self) -> bool:
+        with self._db_lock:
+            if self._db is not None:
+                return True
+            else:
+                self._db = self.open_db()
+                return self._db is not None
```
Comment (tchaikov) [inline src/pybind/mgr/mgr_module.py:1099]: nit,
```suggestion
            try:
                return self.db is not None
            except MgrDBNotReady:
                return False
```

**Review by tchaikov — APPROVED** [review]

File: src/pybind/mgr/mgr_module.py (hunk 8 above)
Comment (batrick) [reply src/pybind/mgr/mgr_module.py:1024]: Thanks, fixed!

File: src/pybind/mgr/mgr_module.py (hunk 9 above)
Comment (batrick) [reply src/pybind/mgr/mgr_module.py]: Thanks, fixed!

**Comment by batrick** [comment]
Fixed conflict.

**Review by tchaikov — APPROVED** [review]

**Review by tchaikov — CHANGES_REQUESTED** [review]
```
2021-04-30T04:57:18.144 DEBUG:teuthology.orchestra.run.smithi032:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd dump --format=json
2021-04-30T04:57:18.407 INFO:teuthology.orchestra.run.smithi032.stdout:
...
"pools":[{"pool":1,"pool_name":".mgr","create_ti
me":"2021-04-30T04:57:17.657219+0000","flags":32769
...
...
2021-04-30T04:58:06.590 INFO:tasks.workunit:Running workunit mon/pg_autoscaler.sh...
2021-04-30T04:58:06.591 DEBUG:teuthology.orchestra.run.smithi032:workunit test mon/pg_autoscaler.sh> mkdir -p -- /home/ubuntu/cephtest/mnt.0/client.0/tmp && cd -- /home/ubuntu/cephtest/mnt.0/client.0/tmp && CEPH_CLI_TEST_DUP_COMMAND=1 CEPH_REF=5fd73215a7021f7c0d11f19efc4f
03c29b8724d9 TESTDIR="/home/ubuntu/cephtest" CEPH_ARGS="--cluster ceph" CEPH_ID="0" PATH=$PATH:/usr/sbin CEPH_BASE=/home/ubuntu/cephtest/clone.client.0 CEPH_ROOT=/home/ubuntu/cephtest/clone.client.0 adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeo
ut 3h /home/ubuntu/cephtest/clone.client.0/qa/workunits/mon/pg_autoscaler.sh
2021-04-30T04:58:06.603 INFO:tasks.workunit.client.0.smithi032.stderr:++ ceph osd ls
2021-04-30T04:58:06.603 INFO:tasks.workunit.client.0.smithi032.stderr:++ wc -l
2021-04-30T04:58:06.958 INFO:tasks.workunit.client.0.smithi032.stderr:+ NUM_OSDS=8
2021-04-30T04:58:06.958 INFO:tasks.workunit.client.0.smithi032.stderr:+ '[' 8 -lt 6 ']'
2021-04-30T04:58:06.959 INFO:tasks.workunit.client.0.smithi032.stderr:++ ceph osd pool ls
2021-04-30T04:58:06.959 INFO:tasks.workunit.client.0.smithi032.stderr:++ wc -l
2021-04-30T04:58:07.304 INFO:tasks.workunit.client.0.smithi032.stderr:+ NUM_POOLS=1
2021-04-30T04:58:07.304 INFO:tasks.workunit.client.0.smithi032.stderr:+ '[' 1 -gt 0 ']'
2021-04-30T04:58:07.305 INFO:tasks.workunit.client.0.smithi032.stderr:+ echo 'test requires no preexisting pools'
2021-04-30T04:58:07.305 INFO:tasks.workunit.client.0.smithi032.stdout:test requires no preexisting pools
2021-04-30T04:58:07.305 DEBUG:teuthology.orchestra.run:got remote process result: 1
```

see https://pulpito.ceph.com/kchai-2021-04-30_03:09:53-rados-wip-kefu-testing-2021-04-30-0020-distro-basic-smithi/

**Comment by batrick** [comment]
@neha-ojha @jdurgin @tchaikov how do you want to address this test failure? One possibility is to delete the .mgr pool but that's hacky and likely to break all sorts of things down the road. The best thing to do is make the test tolerant of an existing (small) pool but not sure how difficult that would be. Any thoughts?

**Comment by jdurgin** [comment]
there are probably a few tests that assume no pools exist, it's probably best to make them work with the .mgr pool existing - @kamoltat can point out a few he's had to change for the autoscaler for instance

**Comment by batrick** [comment]
> there are probably a few tests that assume no pools exist, it's probably best to make them work with the .mgr pool existing - @kamoltat can point out a few he's had to change for the autoscaler for instance

Turns out there's a lot of tests which rely on this. I've added a commit which works around the problem (in the same we do already). PTAL.

File: src/pybind/mgr/mgr_module.py [hunk 10]
```diff
@@ -920,6 +940,176 @@ def cluster_log(self, channel: str, priority: ClusterLogPrio, message: str) -> N
...
+
+    def appify_pool(self, name: str, app: str) -> None:
+        c = {
+            'prefix': 'osd pool application enable',
+            'format': 'json',
+            'pool': name,
+            'app': app,
+            'yes_i_really_mean_it': True
+        }
+        self.check_mon_command(c)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/mgr_module.py:984]: can we keep the names of the arguments equal to the keys of the dicts?

Comment (batrick) [reply src/pybind/mgr/mgr_module.py:984]: Sure, fixed.

**Comment by tchaikov** [comment]
@batrick this 

> > there are probably a few tests that assume no pools exist, it's probably best to make them work with the .mgr pool existing - @kamoltat can point out a few he's had to change for the autoscaler for instance
> 
> Turns out there's a lot of tests which rely on this. I've added a commit which works around the problem (in the same we do already). PTAL.

this is not enough to cure all the failures. some tests assume there is no objects in the cluster. for instance, in https://pulpito.ceph.com/kchai-2021-05-12_08:37:28-rados-wip-kefu-testing-2021-05-12-1357-distro-basic-smithi/6111298/, `qa/standalone/mon/health-mute.sh` checks the healthy status of cluster by looking at the output of `ceph health`.

before your change, we have
```
  cluster:
    id:     d4841920-fe1a-4d03-8c25-f71fb1ca2ce7
    health: HEALTH_OK
            (muted: OSDMAP_FLAGS OSD_DOWN)

  services:
    mon: 1 daemons, quorum a (age 91s)
    mgr: x(active, since 87s)
    osd: 3 osds: 1 up (since 0.369412s), 3 in (since 74s)
         flags noup

  data:
    pools:   1 pools, 8 pgs
    objects: 0 objects, 0 B
    usage:   16 MiB used, 300 GiB / 300 GiB avail
    pgs:     8 active+undersized
```

after your change, we have
```
  cluster:
    id:     b43bf868-b66b-4927-a131-5d19ce01b252
    health: HEALTH_WARN
            Degraded data redundancy: 2/6 objects degraded (33.333%), 1 pg degraded
            (muted: OSDMAP_FLAGS OSD_DOWN)

  services:
    mon: 1 daemons, quorum a (age 78s)
    mgr: x(active, since 73s)
    osd: 3 osds: 2 up (since 10s), 3 in (since 60s)

  data:
    pools:   2 pools, 9 pgs
    objects: 2 objects, 449 KiB
    usage:   18 MiB used, 300 GiB / 300 GiB avail
    pgs:     2/6 objects degraded (33.333%)
             8 active+undersized
             1 active+undersized+degraded
```

i only looked into one of the failures in https://pulpito.ceph.com/kchai-2021-05-12_08:37:28-rados-wip-kefu-testing-2021-05-12-1357-distro-basic-smithi/. i suspect there are more instances of this kind of failure in my batch.

**Comment by batrick** [comment]
performed trivial rebase; going to investigate the failures Kefu shared

**Comment by batrick** [comment]
@tchaikov I had missed a conversion:

https://github.com/ceph/ceph/pull/40740/files#diff-576fa568fce477400315ece595058b8000497a7151b2c5a71f7d0fdd00d87bc2

I think that caused most of the failures. Can you try again?

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-15_10:00:36-rados-wip-kefu-testing-2021-05-15-1545-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-16_04:30:13-rados-wip-kefu-testing-2021-05-16-1043-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-16_11:19:39-rados-wip-kefu-testing-2021-05-16-1043-distro-basic-smithi/

the last set of test suites was tested with master

- https://pulpito.ceph.com/kchai-2021-05-16_11:16:52-rados-master-distro-basic-smithi/

so i don't think this batch was clean. i am dropping this change.

**Comment by tchaikov** [comment]
```
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_cc3bac844fd368a871b19f94cc5d520695b86d44/teuthology/run_tasks.py", line 91, in run_tasks
    manager = run_one_task(taskname, ctx=ctx, config=config)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_cc3bac844fd368a871b19f94cc5d520695b86d44/teuthology/run_tasks.py", line 70, in run_one_task
    return task(**kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_d4a50358e1e0b33978a7d1f185f9bb25f138b1a3/qa/tasks/scrub_test.py", line 390, in task
    test_repair_corrupted_obj(ctx, manager, pg, osd_remote, obj_path, 'rbd')
  File "/home/teuthworker/src/git.ceph.com_ceph-c_d4a50358e1e0b33978a7d1f185f9bb25f138b1a3/qa/tasks/scrub_test.py", line 93, in test_repair_corrupted_obj
    deep_scrub(manager, pg, pool)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_d4a50358e1e0b33978a7d1f185f9bb25f138b1a3/qa/tasks/scrub_test.py", line 78, in deep_scrub
    assert inconsistent
AssertionError
2021-05-16T11:37:00.348 ERROR:teuthology.run_tasks: Sentry event: https://sentry.ceph.com/organizations/ceph/?query=6b874bf060a6406fb1a0e3a2c8bb3dd1
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_cc3bac844fd368a871b19f94cc5d520695b86d44/teuthology/run_tasks.py", line 91, in run_tasks
    manager = run_one_task(taskname, ctx=ctx, config=config)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_cc3bac844fd368a871b19f94cc5d520695b86d44/teuthology/run_tasks.py", line 70, in run_one_task
    return task(**kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_d4a50358e1e0b33978a7d1f185f9bb25f138b1a3/qa/tasks/scrub_test.py", line 390, in task
    test_repair_corrupted_obj(ctx, manager, pg, osd_remote, obj_path, 'rbd')
  File "/home/teuthworker/src/git.ceph.com_ceph-c_d4a50358e1e0b33978a7d1f185f9bb25f138b1a3/qa/tasks/scrub_test.py", line 93, in test_repair_corrupted_obj
    deep_scrub(manager, pg, pool)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_d4a50358e1e0b33978a7d1f185f9bb25f138b1a3/qa/tasks/scrub_test.py", line 78, in deep_scrub
    assert inconsistent
AssertionError
```

i have the failure above in https://pulpito.ceph.com/kchai-2021-05-16_11:19:39-rados-wip-kefu-testing-2021-05-16-1043-distro-basic-smithi/6117396/

**Comment by batrick** [comment]
@tchaikov I fixed `scrub_test` in the latest commit. Here's a rados run for that test:

https://pulpito.ceph.com/pdonnell-2021-06-12_05:51:16-rados:basic-wip-pdonnell-testing-20210612.024127-distro-basic-smithi/

I believe that was the last and only failure attributable to this branch, right? From your posted runs in https://github.com/ceph/ceph/pull/40740#issuecomment-841812446 it seems that this was the only one failure that didn't exist in master.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-06-13_08:30:20-rados-wip-kefu-testing-2021-06-13-1215-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-06-13_13:23:30-rados-wip-kefu-testing-2021-06-13-2001-distro-basic-smithi/

**Review by tchaikov — APPROVED** [review]

**Comment by batrick** [comment]
Wonderful! Thanks for testing @tchaikov.

**Pull request merged**